example/lib/ide/ide_controller.dart

```diff
@@ -0,0 +1,223 @@
+import 'dart:collection';
+import 'dart:io';
+
+import 'package:flutter/foundation.dart';
+
+/// Provides the IDE functionallity, like opening and editing files.
+///
+/// Holds the list of open editors and the active editor.
+class IdeController {
+  /// The list of open editors and the active editor index.
+  ///
+  /// An editor is a view that allows the user to edit a file.
+  ///
+  /// It's separated from the [openFiles] so we can implement, in the future,
+  /// multiple editors for the same file. For example, open two editors side by side.
+  ValueListenable<IdeEditors> get editorData => _editorData;
+  final ValueNotifier<IdeEditors> _editorData = ValueNotifier(IdeEditors(openEditors: []));
+
+  /// The files that have at least one editor open.
+  Set<IdeFile> get openFiles => _openFiles;
+  final Set<IdeFile> _openFiles = {};
+
+  /// Maps the file URI to the opened editors for that file.
+  final _fileUriToOpenedEditors = <String, List<IdeFileEditor>>{};
+
+  /// Maps the file URI to the opened file.
+  final _fileUriToIdeFile = <String, IdeFile>{};
+
+  /// Opens the file with the given URI at the active editor.
+  ///
+  /// If there is no active editor, opens a new editor.
+  ///
+  /// If the file is already open in another editor, shows the existing editor.
+  Future<void> openFileAtActiveEditor(String fileUri) async {
+    if (showExistingEditorForFile(fileUri)) {
+      // The file is already open in another editor. Fizzle.
+      return;
+    }
+
+    await _loadFile(fileUri);
+
+    final activeEditorIndex = _editorData.value.activeEditorIndex;
+    if (activeEditorIndex != null) {
+      // There is an open editor. Close it.
+      closeEditorAtIndex(activeEditorIndex);
+    }
+
+    await _openFile(fileUri, activeEditorIndex ?? 0);
+  }
+
+  /// Opens the file with the given URI at a new editor.
+  ///
+  /// If the file is already open in another editor, shows the existing editor.
+  Future<void> openFileAtNewTab(String fileUri) async {
+    if (showExistingEditorForFile(fileUri)) {
+      return;
+    }
+
+    await _loadFile(fileUri);
+
+    await _openFile(fileUri, _editorData.value.openEditors.length);
+  }
+
+  /// Shows the existing editor for the file with the given URI, if any.
```

Comment:
Please try to aim the explanation at the questions the reader is likely to have. In this case, the obvious question would be "where did the editor go? why would I need to 'show' it? what does that mean?"

Reply:
Updated.